File: lefant/doc-audit/scripts/check-links.ts
#!/usr/bin/env npx ts-node
/**
 * Check markdown files for broken internal links and references.
 *
 * Usage:
 *   npx ts-node check-links.ts [PATH]      # Check path (default: current dir)
 *   npx ts-node check-links.ts --json      # Output as JSON
 *   npx ts-node check-links.ts --fix       # Suggest fixes for broken links
 *
 * Or compile and run:
 *   tsc check-links.ts && node check-links.js [PATH]
 */

import * as fs from "fs";
import * as path from "path";

interface Issue {
  file: string;
  line: number;
  severity: "critical" | "warning" | "info";
  category: string;
  message: string;
  suggestion?: string;
}

function slugify(text: string): string {
  return text
    .toLowerCase()
    .trim()
    .replace(/[^\w\s-]/g, "")
    .replace(/[-\s]+/g, "-");
}

function extractHeadings(content: string): Set<string> {
  const headings = new Set<string>();
  const regex = /^#{1,6}\s+(.+?)(?:\s*\{#([^}]+)\})?$/gm;
  let match;

  while ((match = regex.exec(content)) !== null) {
    const [, text, explicitId] = match;
    if (explicitId) {
      headings.add(explicitId);
    } else {
      headings.add(slugify(text));
    }
  }

  return headings;
}

function extractLinks(
  content: string
): Array<{ line: number; text: string; target: string }> {
  const links: Array<{ line: number; text: string; target: string }> = [];
  const lines = content.split("\n");
  let inCodeBlock = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    // Track code blocks
    if (line.trim().startsWith("```")) {
      inCodeBlock = !inCodeBlock;
      continue;
    }
    if (inCodeBlock) continue;

    // Remove inline code
    const lineNoCode = line.replace(/`[^`]+`/g, "");

    // Markdown links: [text](url)
    const linkRegex = /\[([^\]]*)\]\(([^)]+)\)/g;
    let match;
    while ((match = linkRegex.exec(lineNoCode)) !== null) {
      links.push({ line: i + 1, text: match[1], target: match[2] });
    }

    // Reference-style links: [text][ref]
    const refRegex = /\[([^\]]+)\]\[([^\]]*)\]/g;
    while ((match = refRegex.exec(lineNoCode)) !== null) {
      links.push({
        line: i + 1,
        text: match[1],
        target: `ref:${match[2] || match[1]}`,
      });
    }
  }

  return links;
}

function extractImages(
  content: string
): Array<{ line: number; path: string }> {
  const images: Array<{ line: number; path: string }> = [];
  const lines = content.split("\n");
  let inCodeBlock = false;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];

    if (line.trim().startsWith("```")) {
      inCodeBlock = !inCodeBlock;
      continue;
    }
    if (inCodeBlock) continue;

    const lineNoCode = line.replace(/`[^`]+`/g, "");
    const imgRegex = /!\[[^\]]*\]\(([^)]+)\)/g;
    let match;

    while ((match = imgRegex.exec(lineNoCode)) !== null) {
      images.push({ line: i + 1, path: match[1] });
    }
  }

  return images;
}

function findMarkdownFiles(dir: string): string[] {
  const files: string[] = [];

  function walk(currentDir: string) {
    const entries = fs.readdirSync(currentDir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(currentDir, entry.name);

      // Skip common non-doc directories
      if (entry.isDirectory()) {
        if (
          ["node_modules", ".git", "dist", "build", ".next", "coverage"].includes(
            entry.name
          )
        ) {
          continue;
        }
        walk(fullPath);
      } else if (entry.isFile() && /\.(md|mdx)$/i.test(entry.name)) {
        files.push(fullPath);
      }
    }
  }

  walk(dir);
  return files;
}

function findSimilarFiles(
  target: string,
  searchDir: string,
  maxResults = 3
): string[] {
  const targetName = path.basename(target).toLowerCase();
  const candidates: Array<{ score: number; path: string }> = [];

  try {
    const allFiles = findMarkdownFiles(searchDir);

    for (const f of allFiles) {
      const name = path.basename(f).toLowerCase();
      let score = 0;

      if (targetName.includes(name) || name.includes(targetName)) {
        score += 2;
      }
      if (path.extname(f) === path.extname(target)) {
        score += 1;
      }

      if (score > 0) {
        candidates.push({ score, path: path.relative(searchDir, f) });
      }
    }
  } catch {
    // Ignore errors during suggestion search
  }

  candidates.sort((a, b) => b.score - a.score);
  return candidates.slice(0, maxResults).map((c) => c.path);
}

function checkFile(filePath: string, root: string): Issue[] {
  const issues: Issue[] = [];
  let content: string;

  try {
    content = fs.readFileSync(filePath, "utf-8");
  } catch (e) {
    return [
      {
        file: path.relative(root, filePath),
        line: 0,
        severity: "critical",
        category: "read-error",
        message: `Cannot read file: ${e}`,
      },
    ];
  }

  const fileDir = path.dirname(filePath);
  const headings = extractHeadings(content);

  // Check links
  for (const { line, text, target } of extractLinks(content)) {
    // Skip reference-style links
    if (target.startsWith("ref:")) continue;

    // Skip external links
    if (/^(https?|mailto|ftp):\/\//i.test(target)) continue;

    // Split path and anchor
    const hashIndex = target.indexOf("#");
    const pathPart =
      hashIndex >= 0 ? decodeURIComponent(target.slice(0, hashIndex)) : decodeURIComponent(target);
    const anchor = hashIndex >= 0 ? target.slice(hashIndex + 1) : "";

    if (pathPart) {
      const targetPath = path.resolve(fileDir, pathPart);
      const relTarget = path.relative(root, targetPath);

      if (!fs.existsSync(targetPath)) {
        const similar = findSimilarFiles(pathPart, root);
        issues.push({
          file: path.relative(root, filePath),
          line,
          severity: "critical",
          category: "broken-link",
          message: `Link target not found: ${pathPart}`,
          suggestion: similar.length
            ? `Did you mean: ${similar.join(", ")}`
            : undefined,
        });
      } else if (anchor) {
        // Check anchor in target file
        try {
          const targetContent = fs.readFileSync(targetPath, "utf-8");
          const targetHeadings = extractHeadings(targetContent);

          if (!targetHeadings.has(anchor)) {
            const available = Array.from(targetHeadings).slice(0, 5).join(", ");
            issues.push({
              file: path.relative(root, filePath),
              line,
              severity: "warning",
              category: "missing-anchor",
              message: `Anchor #${anchor} not found in ${pathPart}`,
              suggestion: targetHeadings.size
                ? `Available anchors: ${available}`
                : undefined,
            });
          }
        } catch {
          // Ignore read errors for anchor check
        }
      }
    } else if (anchor) {
      // Same-file anchor
      if (!headings.has(anchor)) {
        const available = Array.from(headings).slice(0, 5).join(", ");
        issues.push({
          file: path.relative(root, filePath),
          line,
          severity: "warning",
          category: "missing-anchor",
          message: `Anchor #${anchor} not found in this file`,
          suggestion: headings.size ? `Available anchors: ${available}` : undefined,
        });
      }
    }
  }

  // Check images
  for (const { line, path: imgPath } of extractImages(content)) {
    if (/^(https?|data):/i.test(imgPath)) continue;

    const targetPath = path.resolve(fileDir, decodeURIComponent(imgPath));

    if (!fs.existsSync(targetPath)) {
      issues.push({
        file: path.relative(root, filePath),
        line,
        severity: "critical",
        category: "missing-image",
        message: `Image not found: ${imgPath}`,
      });
    }
  }

  return issues;
}

function main() {
  const args = process.argv.slice(2);
  const jsonOutput = args.includes("--json");
  const showFix = args.includes("--fix");
  const targetPath = args.find((a) => !a.startsWith("--")) || ".";

  const root = path.resolve(targetPath);

  if (!fs.existsSync(root)) {
    console.error(`Error: Path not found: ${root}`);
    process.exit(1);
  }

  // Find markdown files
  let mdFiles: string[];
  if (fs.statSync(root).isFile()) {
    mdFiles = [root];
  } else {
    mdFiles = findMarkdownFiles(root);
  }

  // Check all files
  const allIssues: Issue[] = [];
  for (const f of mdFiles) {
    const issues = checkFile(f, fs.statSync(root).isFile() ? path.dirname(root) : root);
    allIssues.push(...issues);
  }

  // Output
  if (jsonOutput) {
    console.log(JSON.stringify(allIssues, null, 2));
  } else {
    if (allIssues.length === 0) {
      console.log(`No issues found in ${mdFiles.length} files.`);
      process.exit(0);
    }

    const critical = allIssues.filter((i) => i.severity === "critical");
    const warnings = allIssues.filter((i) => i.severity === "warning");
    const info = allIssues.filter((i) => i.severity === "info");

    console.log(`Found ${allIssues.length} issues in ${mdFiles.length} files\n`);

    const groups: Array<[string, Issue[]]> = [
      ["CRITICAL", critical],
      ["WARNING", warnings],
      ["INFO", info],
    ];

    for (const [label, issues] of groups) {
      if (issues.length === 0) continue;

      console.log(`## ${label} (${issues.length})\n`);

      for (const i of issues) {
        console.log(`  ${i.file}:${i.line} [${i.category}]`);
        console.log(`    ${i.message}`);
        if (showFix && i.suggestion) {
          console.log(`    -> ${i.suggestion}`);
        }
        console.log();
      }
    }
  }

  process.exit(allIssues.length > 0 ? 1 : 0);
}

main();

File: lefant/doc-audit/scripts/check-links.py
#!/usr/bin/env python3
"""
Check markdown files for broken internal links and references.

Usage:
    python check-links.py [PATH]         # Check path (default: current dir)
    python check-links.py --json         # Output as JSON
    python check-links.py --fix          # Suggest fixes for broken links
"""

import argparse
import json
import os
import re
import sys
from pathlib import Path
from typing import NamedTuple
from urllib.parse import unquote, urlparse


class Issue(NamedTuple):
    file: str
    line: int
    severity: str  # critical, warning, info
    category: str  # broken-link, missing-anchor, missing-image, etc.
    message: str
    suggestion: str | None = None


def slugify(text: str) -> str:
    """Convert heading text to GitHub-style anchor slug."""
    text = text.lower().strip()
    text = re.sub(r'[^\w\s-]', '', text)
    text = re.sub(r'[-\s]+', '-', text)
    return text


def extract_headings(content: str) -> set[str]:
    """Extract all heading anchors from markdown content."""
    headings = set()
    for match in re.finditer(r'^#{1,6}\s+(.+?)(?:\s*{#([^}]+)})?$', content, re.MULTILINE):
        text, explicit_id = match.groups()
        if explicit_id:
            headings.add(explicit_id)
        else:
            headings.add(slugify(text))
    return headings


def extract_links(content: str, file_path: Path) -> list[tuple[int, str, str]]:
    """Extract all links from markdown: (line_number, link_text, link_target)."""
    links = []
    in_code_block = False

    for i, line in enumerate(content.split('\n'), 1):
        # Track code blocks
        if line.strip().startswith('```'):
            in_code_block = not in_code_block
            continue
        if in_code_block:
            continue

        # Skip inline code
        line_no_code = re.sub(r'`[^`]+`', '', line)

        # Markdown links: [text](url)
        for match in re.finditer(r'\[([^\]]*)\]\(([^)]+)\)', line_no_code):
            links.append((i, match.group(1), match.group(2)))
        # Reference-style links: [text][ref] or [ref]
        for match in re.finditer(r'\[([^\]]+)\]\[([^\]]*)\]', line_no_code):
            links.append((i, match.group(1), f'ref:{match.group(2) or match.group(1)}'))
    return links


def extract_images(content: str) -> list[tuple[int, str]]:
    """Extract all image references: (line_number, image_path)."""
    images = []
    in_code_block = False

    for i, line in enumerate(content.split('\n'), 1):
        if line.strip().startswith('```'):
            in_code_block = not in_code_block
            continue
        if in_code_block:
            continue

        # Skip inline code
        line_no_code = re.sub(r'`[^`]+`', '', line)

        for match in re.finditer(r'!\[[^\]]*\]\(([^)]+)\)', line_no_code):
            images.append((i, match.group(1)))
    return images


def find_similar_files(target: str, search_dir: Path, max_results: int = 3) -> list[str]:
    """Find files with similar names for suggestions."""
    target_name = Path(target).name.lower()
    candidates = []
    for f in search_dir.rglob('*'):
        if f.is_file() and f.suffix in ('.md', '.mdx', '.rst', '.txt'):
            score = 0
            name = f.name.lower()
            if target_name in name or name in target_name:
                score += 2
            if f.suffix == Path(target).suffix:
                score += 1
            if score > 0:
                candidates.append((score, str(f.relative_to(search_dir))))
    candidates.sort(reverse=True)
    return [c[1] for c in candidates[:max_results]]


def check_file(file_path: Path, all_files: dict[str, Path], root: Path) -> list[Issue]:
    """Check a single markdown file for issues."""
    issues = []
    try:
        content = file_path.read_text(encoding='utf-8')
    except Exception as e:
        return [Issue(str(file_path), 0, 'critical', 'read-error', f'Cannot read file: {e}')]

    file_dir = file_path.parent
    headings = extract_headings(content)

    # Check links
    for line_num, text, target in extract_links(content, file_path):
        # Skip reference-style links (would need separate pass)
        if target.startswith('ref:'):
            continue

        # Skip external links
        parsed = urlparse(target)
        if parsed.scheme in ('http', 'https', 'mailto', 'ftp'):
            continue

        # Split path and anchor
        path_part, _, anchor = unquote(target).partition('#')

        if path_part:
            # Resolve relative path
            target_path = (file_dir / path_part).resolve()
            try:
                rel_target = target_path.relative_to(root)
            except ValueError:
                rel_target = target_path

            if not target_path.exists():
                similar = find_similar_files(path_part, root)
                suggestion = f"Did you mean: {', '.join(similar)}" if similar else None
                issues.append(Issue(
                    str(file_path.relative_to(root)),
                    line_num,
                    'critical',
                    'broken-link',
                    f"Link target not found: {path_part}",
                    suggestion
                ))
            elif anchor:
                # Check anchor in target file
                try:
                    target_content = target_path.read_text(encoding='utf-8')
                    target_headings = extract_headings(target_content)
                    if anchor not in target_headings:
                        issues.append(Issue(
                            str(file_path.relative_to(root)),
                            line_num,
                            'warning',
                            'missing-anchor',
                            f"Anchor #{anchor} not found in {path_part}",
                            f"Available anchors: {', '.join(sorted(target_headings)[:5])}" if target_headings else None
                        ))
                except Exception:
                    pass
        elif anchor:
            # Same-file anchor
            if anchor not in headings:
                issues.append(Issue(
                    str(file_path.relative_to(root)),
                    line_num,
                    'warning',
                    'missing-anchor',
                    f"Anchor #{anchor} not found in this file",
                    f"Available anchors: {', '.join(sorted(headings)[:5])}" if headings else None
                ))

    # Check images
    for line_num, img_path in extract_images(content):
        parsed = urlparse(img_path)
        if parsed.scheme in ('http', 'https', 'data'):
            continue

        target_path = (file_dir / unquote(img_path)).resolve()
        if not target_path.exists():
            issues.append(Issue(
                str(file_path.relative_to(root)),
                line_num,
                'critical',
                'missing-image',
                f"Image not found: {img_path}"
            ))

    return issues


def main():
    parser = argparse.ArgumentParser(description='Check markdown files for broken links')
    parser.add_argument('path', nargs='?', default='.', help='Path to check')
    parser.add_argument('--json', action='store_true', help='Output as JSON')
    parser.add_argument('--fix', action='store_true', help='Show fix suggestions')
    args = parser.parse_args()

    root = Path(args.path).resolve()
    if not root.exists():
        print(f"Error: Path not found: {root}", file=sys.stderr)
        sys.exit(1)

    # Find all markdown files
    if root.is_file():
        md_files = [root]
        root = root.parent
    else:
        md_files = list(root.rglob('*.md')) + list(root.rglob('*.mdx'))

    # Build file index
    all_files = {str(f.relative_to(root)): f for f in md_files}

    # Check all files
    all_issues: list[Issue] = []
    for f in md_files:
        issues = check_file(f, all_files, root)
        all_issues.extend(issues)

    # Output
    if args.json:
        print(json.dumps([i._asdict() for i in all_issues], indent=2))
    else:
        if not all_issues:
            print(f"No issues found in {len(md_files)} files.")
            sys.exit(0)

        # Group by severity
        critical = [i for i in all_issues if i.severity == 'critical']
        warnings = [i for i in all_issues if i.severity == 'warning']
        info = [i for i in all_issues if i.severity == 'info']

        print(f"Found {len(all_issues)} issues in {len(md_files)} files\n")

        for label, issues in [('CRITICAL', critical), ('WARNING', warnings), ('INFO', info)]:
            if not issues:
                continue
            print(f"## {label} ({len(issues)})\n")
            for i in issues:
                print(f"  {i.file}:{i.line} [{i.category}]")
                print(f"    {i.message}")
                if args.fix and i.suggestion:
                    print(f"    -> {i.suggestion}")
                print()

    sys.exit(1 if all_issues else 0)


if __name__ == '__main__':
    main()
